### PR TITLE
[Feature] enable socket net_type for rpc

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

`socket` is enabled back which was the main rpc backend in `0.7.x`. This is supposed to fix the hang issue mentioned in https://github.com/dmlc/dgl/issues/3881.

Try with below command and it works well with `net_type='socket'`:
`DGL_HOME=/home/ubuntu/workspace/dgl_p0 DGL_LIBRARY_PATH=$DGL_HOME/build PYTHONPATH=.:$DGL_HOME/python:$PYTHONPATH python3 ~/workspace/dgl_p0/tools/launch.py --workspace ~/workspace/dgl_p0/examples/pytorch/graphsage/dist/ --num_trainers 4 --num_samplers 4 --num_servers 4 --part_config data/ogb-product.json --ip_config ~/workspace/ip_config.txt " python3 train_dist.py --graph_name ogb-product --ip_config ~/workspace/ip_config.txt --num_epochs 5 --batch_size 256 --eval_every 5"`
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by zheng-da** [comment]
is this the same as reverting the code to use the socket for communication?

**Comment by Rhett-Ying** [comment]
Almost, though not exactly the same.

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: python/dgl/distributed/rpc_client.py [hunk 1]
```diff
@@ -222,7 +226,7 @@ def shutdown_servers(ip_config, num_servers):
     rpc.register_sig_handler()
     server_namebook = rpc.read_ip_config(ip_config, num_servers)
     num_servers = len(server_namebook)
-    rpc.create_sender(MAX_QUEUE_SIZE, 'socket')
+    rpc.create_sender(MAX_QUEUE_SIZE, 'tensorpipe')
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply python/dgl/distributed/rpc_client.py:229]: I just realized this cannot be checked as this function is called in a separate process which always expect server is in `tensorpipe` and cannot record or know the server's backend. So I'd like to leave it here.

File: python/dgl/distributed/rpc_server.py [hunk 2]
```diff
@@ -83,11 +88,20 @@ def start_server(server_id, ip_config, num_servers, num_clients, server_state, \
...
             for client_id, addr in client_namebook.items():
                 client_ip, client_port = addr.split(':')
-                # TODO[Rhett]: server should not be blocked endlessly.
+                try_times = 0
                 while not rpc.connect_receiver(client_ip, client_port, client_id, group_id):
+                    try_times += 1
+                    if try_times >= max_try_times:
+                        raise DGLError("Failed to connect receiver[{}:{}] after retry {} times. "
+                                       "Please check availability of target receiver or change max "
+                                       "try times via env 'DGL_DIST_MAX_TRY_TIMES'.".format(
```
Comment (jermainewang) [inline python/dgl/distributed/rpc_server.py:101]: Failed to connect to receiver [{}:{}] after {} retries. Please check availability of this target receiver or change the max retry times via 'DGL_DIST_MAX_TRY_TIMES'.

**Pull request merged**